### PR TITLE
Refactor app-preferences-registry to use di

### PR DESCRIPTION
**Pull request by jim-docker**

### CODE REVIEW

File: src/renderer/components/+preferences/preferences.tsx [hunk 1]
```diff
@@ -50,18 +49,18 @@ import { LensProxy } from "./proxy";
 import { Telemetry } from "./telemetry";
 import { Extensions } from "./extensions";
 import { sentryDsn } from "../../../common/vars";
+import { withInjectables } from "@ogre-tools/injectable-react";
+import type { AppPreferenceRegistration } from "./app-preferences/app-preference-registration";
+import appPreferencesInjectable from "./app-preferences/app-preferences.injectable";
 
-@observer
-export class Preferences extends React.Component {
-  @observable historyLength: number | undefined;
```
Comment (jim-docker) [inline src/renderer/components/+preferences/preferences.tsx:40]: This was never used, unless I missed something

File: src/renderer/components/+preferences/app-preferences/get-app-preferences.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+
+export const getAppPreferences = ({ extensions }: Dependencies) => {
+  return computed(() => {
+    return [
+      ...extensions.get()
+        .flatMap((extension) => extension.appPreferences)
+        .map(pref => getRegisteredItem(pref)),
+    ];},
+  );
```
Comment (Nokel81) [inline src/renderer/components/+preferences/app-preferences/get-app-preferences.ts]: nit, there is never a need to spread a `.map()` call.

```suggestion
  return computed(() => (
    extensions.get()
      .flatMap((extension) => extension.appPreferences)
      .map(getRegisteredItem)
  ));
```

File: src/renderer/components/+preferences/app-preferences/app-preference-registration.d.ts [hunk 3]
```diff
@@ -33,16 +32,3 @@ export interface AppPreferenceRegistration {
   showInPreferencesTab?: string;
   components: AppPreferenceComponents;
 }
-
-export interface RegisteredAppPreference extends AppPreferenceRegistration {
-  id: string;
-}
```
Comment (Nokel81) [inline src/renderer/components/+preferences/app-preferences/app-preference-registration.d.ts]: I don't think we should remove this type. I think that the injectable's type should be `IComputedValue<RegisteredAppPreference[]>`.

That way we communicate that `id` is always defined (which is good when we (hopefully soon) turn on `strict`)

File: src/renderer/components/+preferences/application.tsx [hunk 4]
```diff
@@ -56,7 +62,7 @@ export const Application = observer(() => {
 
   const [customUrl, setCustomUrl] = React.useState(userStore.extensionRegistryUrl.customUrl || "");
   const [shell, setShell] = React.useState(userStore.shell || "");
-  const extensionSettings = AppPreferenceRegistry.getInstance().getItems().filter((preference) => preference.showInPreferencesTab === "application");
+  const extensionSettings = appPreferenceItems.get().filter((preference) => preference.showInPreferencesTab === "application");
```
Comment (Nokel81) [inline src/renderer/components/+preferences/application.tsx:49]: To go even further, there could be an injectable whose type is

```
IComputedValue<Map<string, RegisteredAppPreference[]>>
```

which does the splitting once

File: src/renderer/components/+preferences/extensions.tsx [hunk 5]
```diff
@@ -19,13 +19,21 @@
...
+import appPreferencesInjectable from "./app-preferences/app-preferences.injectable";
 import { ExtensionSettings } from "./extension-settings";
 
-export const Extensions = observer(() => {
-  const settings = AppPreferenceRegistry.getInstance().getItems();
+interface Dependencies {
+  appPreferenceItems: IComputedValue<AppPreferenceRegistration[]>
+}
+
+const NonInjectedExtensions: React.FC<Dependencies> = ({ appPreferenceItems }) => {
```
Comment (Nokel81) [inline src/renderer/components/+preferences/extensions.tsx:18]: This needs to be `observer`. Which also means you can get rid of the type annotation

```
const NonInjectedExtensions = observer(({ appPreferenceItems }: Dependencies) => { ...
```

Comment (Nokel81) [reply src/renderer/components/+preferences/extensions.tsx:18]: Ah I see that you do that below...

File: src/renderer/components/+preferences/app-preferences/get-app-preferences.ts (hunk 2 above)
Comment (jim-docker) [reply src/renderer/components/+preferences/app-preferences/get-app-preferences.ts]: good catch. I was duplicating what was done in https://github.com/lensapp/lens/pull/4615, but there they have another element in the `[ ]`

File: src/renderer/components/+preferences/app-preferences/app-preference-registration.d.ts (hunk 3 above)
Comment (jim-docker) [reply src/renderer/components/+preferences/app-preferences/app-preference-registration.d.ts]: Thanks, I wasn't sure the best way to handle this type. Internally it seems we always use `RegisteredAppPreference`

File: src/renderer/components/+preferences/extensions.tsx (hunk 5 above)
Comment (jim-docker) [reply src/renderer/components/+preferences/extensions.tsx:18]: again, I took this pattern from #4615

Comment (Nokel81) [reply src/renderer/components/+preferences/extensions.tsx:18]: Fair, enough I don't think it matters that much.

File: src/renderer/components/+preferences/telemetry.tsx [hunk 6]
```diff
@@ -22,13 +22,20 @@ import { observer } from "mobx-react";
...
 
-export const Telemetry = observer(() => {
-  const extensions = AppPreferenceRegistry.getInstance().getItems();
+interface Dependencies {
+  appPreferenceItems: IComputedValue<RegisteredAppPreference[]>
+}
+
+const NonInjectedTelemetry: React.FC<Dependencies> = ({ appPreferenceItems }) => {
+  const extensions = appPreferenceItems.get();
   const telemetryExtensions = extensions.filter(e => e.showInPreferencesTab == "telemetry");
```
Comment (Nokel81) [inline src/renderer/components/+preferences/telemetry.tsx:23]: I sort of think that this should be its own injectable

File: src/renderer/components/+preferences/application.tsx (hunk 4 above)
Comment (jim-docker) [reply src/renderer/components/+preferences/application.tsx:49]: I don't follow...

Comment (Nokel81) [reply src/renderer/components/+preferences/application.tsx:49]: We currently have one injectable: `appPreferenceItemsInjectable`. But we could change its type to from `IComputedValue<RegisteredAppPreference[]>` to `IComputedValue<Map<string, RegisteredAppPreference[]>>` and then have an injectable for each of the pages, say called `telemetryPreferenceItemsInjectables`. That way when we come to test each component, we only need to override the single injectable and the filtering is moved out of the component.

Comment (jim-docker) [reply src/renderer/components/+preferences/application.tsx:49]: Sounds good. But that may be better done in another PR? After https://github.com/lensapp/lens/pull/4651 is merged, which is waiting on this PR.

have to resolve conflicts first...

Comment (Nokel81) [reply src/renderer/components/+preferences/application.tsx:49]: Okay sounds good

File: src/renderer/components/+preferences/telemetry.tsx (hunk 6 above)
Comment (aleksfront) [reply src/renderer/components/+preferences/telemetry.tsx:23]: I can do this in #4651

**Review by aleksfront — APPROVED** [review]
LGTM

**Review by Nokel81 — APPROVED** [review]

**Comment by jansav** [comment]
Merging this closed #4649

**Pull request merged**